File: src/main/java/net/Chidoziealways/everythingjapanese/mixins/GuiMixin.java
package net.Chidoziealways.everythingjapanese.mixins;

import net.Chidoziealways.everythingjapanese.capabilities.ModCapabilities;
import net.Chidoziealways.everythingjapanese.chakra.ChakraHUDOverlay;
import net.minecraft.client.DeltaTracker;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.gui.GuiGraphics;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.spongepowered.asm.mixin.Final;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.Unique;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

@Mixin(
        value = {Gui.class},
        priority = 6000
)
public abstract class GuiMixin {

    private static final Logger log = LoggerFactory.getLogger(GuiMixin.class);
    @Shadow @Final private Minecraft minecraft;

    @Inject(method = "renderHotbarAndDecorations", at = @At("TAIL"))
    private void onRender(GuiGraphics pGuiGraphics, DeltaTracker pDeltaTracker, CallbackInfo ci) {
        if (minecraft.player == null) return;
        // Fetch the client-side stored Chakra (updated via packets)
        minecraft.player.getCapability(ModCapabilities.CHAKRA_CAPABILITY).ifPresent(iChakra -> ChakraHUDOverlay.renderChakraBar(pGuiGraphics, iChakra.getChakra(), iChakra.getMaxChakra()));
    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/mixins/GuiMixin.kt
package net.Chidoziealways.everythingjapanese.mixins

import net.Chidoziealways.everythingjapanese.capabilities.ModCapabilities
import net.Chidoziealways.everythingjapanese.chakra.ChakraHUDOverlay.renderChakraBar
import net.Chidoziealways.everythingjapanese.chakra.IChakra
import net.minecraft.client.DeltaTracker
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Gui
import net.minecraft.client.gui.GuiGraphics
import net.minecraftforge.common.util.NonNullConsumer
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.spongepowered.asm.mixin.Final
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

@Mixin(value = [Gui::class], priority = 6000)
abstract class GuiMixin {
    @Shadow
    @Final
    private val minecraft: Minecraft? = null

    @Inject(method = ["renderHotbarAndDecorations"], at = [At("TAIL")])
    private fun onRender(pGuiGraphics: GuiGraphics, pDeltaTracker: DeltaTracker?, ci: CallbackInfo?) {
        if (minecraft!!.player == null) return
        // Fetch the client-side stored Chakra (updated via packets)
        minecraft.player!!.getCapability<IChakra?>(ModCapabilities.CHAKRA_CAPABILITY!!)
            .ifPresent(NonNullConsumer { iChakra: IChakra? ->
                renderChakraBar(
                    pGuiGraphics,
                    iChakra!!.chakra,
                    iChakra.getMaxChakra()
                )
            })
    }

    private companion object {
        private val log: Logger? = LoggerFactory.getLogger(GuiMixin::class.java)
    }
}
